jongbum/Assignment07/src/main/java/com/likelion/basecode/movie/application/MovieService.java

```diff
@@ -0,0 +1,108 @@
+import com.likelion.basecode.movie.api.dto.response.MovieResponseDto;
+import com.likelion.basecode.post.domain.Post;
+import com.likelion.basecode.post.domain.repository.PostRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor
+public class MovieService {
+
+    private final PostRepository postRepository;
+    private final TagRecommendationClient tagClient;
+    private final MovieSearchClient movieSearchClient;
+
+    // 일별 박스오피스 전체 영화 목록 조회 (장르 정보 포함)
+    public MovieListResponseDto fetchAllDailyBoxOfficeMoviesWithGenres() {
+        String targetDate = LocalDate.now().minusDays(1).format(DateTimeFormatter.ofPattern("yyyyMMdd"));
+
+        // 일별 박스오피스 기본 목록 조회 (movieCd 포함)
+        List<MovieResponseDto> moviesWithoutGenres = movieSearchClient.fetchDailyBoxOffice(targetDate);
+
+        // 각 영화에 대해 상세 정보 API 호출하여 장르 정보 보강
+        List<MovieResponseDto> moviesWithGenres = moviesWithoutGenres.stream()
+                .map(movie -> {
+                    // movieCd가 없으면 장르를 가져올 수 없기 때문에
+                    // 이런 경우 발생하면 에러 처리 또는 스킵
+                    if (movie.movieCd() == null || movie.movieCd().isEmpty()) {
+                        System.err.println("MovieCd not found for movie: " + movie.movieNm());
+                        return movie; // movieCd가 없으면 장르 없이 기존 DTO 반환
+                    }
+                    try {
+                        MovieInfoResponseDto movieDetail = movieSearchClient.fetchMovieDetail(movie.movieCd());
+                        // 새로운 MovieResponseDto 생성
+                        return new MovieResponseDto(
+                                movie.movieCd(),
+                                movie.movieNm(),
+                                movie.openDt(),
+                                movie.repNationNm(),
+                                movieDetail.genre() // MovieInfoResponseDto에서 가져온 장르로 채움
+                        );
+                    } catch (BusinessException e) {
+                        System.err.println("Failed to fetch movie detail for movieCd: " + movie.movieCd() + " - " + e.getMessage());
+                        return movie; // 상세 정보 조회 실패 시, 장르 없이 기존 DTO 반환
+                    }
+                })
+                .collect(Collectors.toList());
+
+        return new MovieListResponseDto(moviesWithGenres);
+    }
+
+    // 기존 함수: 특정 게시글의 추천 태그를 기반으로 영화 추천 (이 함수도 변경된 DTO 사용)
+    public MovieListResponseDto recommendMoviesByPostId(Long postId) {
+        Post post = postRepository.findById(postId)
+                .orElseThrow(() -> new BusinessException(ErrorCode.POST_NOT_FOUND_EXCEPTION,
+                        ErrorCode.POST_NOT_FOUND_EXCEPTION.getMessage()));
+
+        List<String> tags = tagClient.getRecommendedTags(post.getContents());
```

Comment:
이 api로 요청을 보내보면 '추천 가능한 태그가 없습니다.' 라고 뜨더라고요. 그래서 추천이 잘 안 된다고 하신 것 같은데 이는 tagClient.getRecoomendedTags()가 빈 리스트를 반환하기에 발생하는 예외인 것 같습니다.
그래서 게시글의 태그를 tagClient.getRecommenedTags()로 다시 요청을 보내 생성하는 것보다, 게시글 생성시 태그는 자동 생성되니 데이터베이스에 저장된 태그를 가져와 해당 태그에 맞는 영화를 추천해 주면 어떨까요 ??

아래와 같이 바꿔주면 태그에 '드래곤'이 있다면 해당 태그가 포함된 영화를 응답으로 보내주기는 합니다 ! 
```suggestion
        Post post = postRepository.findByIdWithTags(postId)
                .orElseThrow(() -> new BusinessException(ErrorCode.POST_NOT_FOUND_EXCEPTION,
                        ErrorCode.POST_NOT_FOUND_EXCEPTION.getMessage()));

        List<String> tags = post.getPostTags().stream()
                .map(postTag -> postTag.getTag().getName())
                .toList();
```

![image](https://github.com/user-attachments/assets/b8880bac-34ee-4365-a02f-371edb74e4ed)

Reply:
오 수정하니까 되는 것 확인했습니다! 감사합니다!!
![추천](https://github.com/user-attachments/assets/400dbe55-b65e-49c0-a25c-756d64baeb77)